### ISSUE TITLE
ContainerCluster nodeConfig immutable/rerun not possible

### ISSUE THREAD

**Issue by Atoms** [GoogleCloudPlatform/k8s-config-connector#441]
## Checklist

- [x] I did not find a related open issue.
- [x] I did not find a solution in the troubleshooting guide: (https://cloud.google.com/config-connector/docs/troubleshooting)

## Bug Description

We have no nat by default for Instances in all subnets, except if they have the tag "internet".  When you configure Kubernetes cluster without nodeConfig parameter in spec then initial nodes cannot finish deploying cause they cannot get to storage.googleapis.com (google private access)
If we add nodeConfig in YAML then we cannot apply once more that manifest, cause nodeConfig is an immutable field.

As we use Ansible for configuring config-connector resources, we will run multiple times playbooks when needed adjustments in fields that are changeable. 

This is blocking our further development and Kubernetes cluster deployment. 
### Kubernetes Cluster Version
Client Version: v1.20.4
Server Version: v1.19.7-gke.2503

### Config Connector Version
1.38.1

### Config Connector Mode
cluster

## Log Output

```
'Failed to patch object: b''{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"admission webhook \\"deny-immutable-field-updates.cnrm.cloud.google.com\\" denied the request: cannot make changes on immutable fields: [nodeConfig]","code":400}
```
### Steps to reproduce the issue

Try this YAML to apply multiple times.
### YAML snippets

```yaml
---
apiVersion: container.cnrm.cloud.google.com/v1beta1
kind: ContainerCluster
metadata:
  name: "cluster-namer"
  annotations:
    cnrm.cloud.google.com/remove-default-node-pool: "true"
spec:
  location: "europe-west3-a"
  initialNodeCount: 1
  defaultMaxPodsPerNode: 60
  nodeLocations: 
    - "europe-west3-b"
  masterAuth:
    clientCertificateConfig:
      issueClientCertificate: true
  nodeConfig:
    tags:
      - internet
  masterAuthorizedNetworksConfig:
    cidrBlocks:
      - cidrBlock: 10.2.0.0/24
        displayName: "gcp-jumpbox"
  privateClusterConfig:
    enablePrivateEndpoint: true
    enablePrivateNodes: true
    masterIpv4CidrBlock: "172.16.0.0/28"
    masterGlobalAccessConfig:
      enabled: false
  networkingMode: "VPC_NATIVE"
  networkRef:
    name: "default-vpc"
  subnetworkRef:
    name: "k8s"
  ipAllocationPolicy:
    servicesSecondaryRangeName: "k8s-services"
    clusterSecondaryRangeName: "k8s-pods"
  releaseChannel:
    channel: STABLE
  enableBinaryAuthorization: false
  enableIntranodeVisibility: false
  enableShieldedNodes: false
  addonsConfig:
    networkPolicyConfig:
      disabled: false
    dnsCacheConfig:
      enabled: true
    configConnectorConfig:
      enabled: false
  networkPolicy:
    enabled: false
  podSecurityPolicyConfig:
    enabled: true
  verticalPodAutoscaling:
    enabled: true
```

**Comment by Atoms** [comment]
Found workaround, but it's ugly in my opinion, as we need to route those Google subnets to default gateway I add 124 subnets (difference between https://www.gstatic.com/ipranges/cloud.json and https://www.gstatic.com/ipranges/goog.json as written in documentation - https://cloud.google.com/vpc/docs/access-apis-external-ip#ip-addr-defaults), but as those lists are dynamic, there is the possibility that I need to rerun task which keeps routes up2date all the time.

**Comment by caieo** [comment]
Hi @Atoms, we're sorry you ran into this! I'm not an expert with cluster routing/subnetworks, but can you clarify if I'm interpreting your reproduction steps correctly?  
1) create a `ContainerCluster` resource without `spec.nodeConfig`
2) add the `nodeConfig` field in  
3) run into an error stating that `nodeConfig` is immutable and thus you are unable to add the necessary tags for your subnets

I took a brief look at the [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/projects.zones.clusters), and it seems like `nodeConfig` is a deprecated field in favor of `nodePool.config`. I think you might be able to achieve a mutable `nodeConfig` by using our `ContainerNodePool` resource and setting your tags there. Here's our documentation for an example of how to use `ContainerNodePool` with a `ContainerCluster`: https://cloud.google.com/config-connector/docs/reference/resource-docs/container/containernodepool#sample_yamls  
Let me know if this workaround does/doesn't work for you.

**Comment by Atoms** [comment]
HI @caieo thanks for trying to understand my issue, and sorry for my bad description about it, I will try again with steps on how to reproduce:
1) Create Cloud nat with a default route to the default internet gateway just for VM's which have network tag "internet"
2) Create YAML with `nodeConfig: tags: ['internet']`
3) Apply YAML, cluster get's created
4) Apply YAML again (e.g. with changes in mutable fields, or no changes at all) you run into an error which states nodeConfig is immutable.

I'm using `ContainerNodePool`, but looking when you look at the documentation for `ContainerCluster`, you see that you need to provide 1 node as initalNodeCount (or I'm not understanding correctly all that)

I create cluster with `initialNodeCount: 1`, when the cluster is ready, annotation `cnrm.cloud.google.com/remove-default-node-pool: "true"` removed that node. and then the next task adds `ContainerNodePool` and attaches that to the cluster. If I don't add `nodeConfig` to `ContainerCluster` resource, then my 1 initial node cannot connect to storage.googleapis.com cause we deny nat for all Instances which do not have the network tag "internet" 

With ansible I'm creating resources, I always wait until resource is ready, so in case adding NodePool first and then cluster to which attach that nodepool, would be not ok for me, cause Nodepool will not be ready until the cluster is not created.

**Comment by caieo** [comment]
Hi @Atoms, thanks for clarifying your reproduction steps for me. 

In your reproduction steps,
> 2. Create YAML with `nodeConfig: tags: ['internet']`

Is the 'YAML' you're creating `ContainerNodePool` or `ContainerCluster`? If I understand your configuration correctly, you should be setting the nodeConfig in `ContainerNodePool` as those are the nodes that the cluster will attach to. 

> If I don't add nodeConfig to ContainerCluster resource, then my 1 initial node cannot connect to storage.googleapis.com cause we deny nat for all Instances which do not have the network tag "internet"

When using the `cnrm.cloud.google.com/remove-default-node-pool: "true"` annotation, the default/initial node created during cluster creation is deleted and instead the cluster will use the specified `ContainerNodePools`. With that in mind, I think your initial node shouldn't exist, let alone connect to storage.googleapis.com. I'm also not too clear what "denying nat for all instances w/o the network tag" entails. If possible, can you share your YAML configurations?

**Comment by Atoms** [comment]
Hi @caieo 

> 2. Create YAML with `nodeConfig: tags: ['internet']`

I use this in `ContainerCluster` as per documentation you cannot create `ContainerCluster` without nodes, therefore we need to create it with one node which is removed afterward and `ContainerNodePool` is attached to this `ContainerCluster`. 

>initialNodeCount: Immutable. The number of nodes to create in this cluster's default node pool. In regional or multi-zonal clusters, this is the number of nodes per zone. Must be set if node_pool is not set. If you're using google_container_node_pool objects with no default node pool, you'll need to set this to a value of at least 1, alongside setting remove_default_node_pool to true.

**Must be set if node_pool is not set.** - there is no setting for node_pool (or I don't understand this properly)

**If you're using google_container_node_pool objects with no default node pool, you'll need to set this to a value of at least 1, alongside setting remove_default_node_pool to true.** - This is what I'm doing as I cannot see how to attach google_container_node_pool  object when creating a cluster.


> I'm also not too clear what "denying nat for all instances w/o the network tag" entails. If possible, can you share your YAML configurations?

```YAML
---
apiVersion: compute.cnrm.cloud.google.com/v1beta1
kind: ComputeRoute
metadata:
  name: "internet-nat-route"
spec:
  destRange: "0.0.0.0/0"
  networkRef:
    name: "default-vpc"
  priority: 1000
  nextHopGateway: "default-internet-gateway"
  tags:
    - "internet"
```
With this `ComputeRoute`, I deny internet access for instances without network tag `internet`, which means also Private Google Access is blocked, so when I create `ContainerCluster` i need to have that one node to have internet (Private Google Access so i need to add tag `internet` to that node. If I don't do that then the node is not ready, and the cluster is also not becoming ready, so I cannot move forward with my playbook to attack `ContainerNodePool` to it. 

Maybe the best way is to add routes for Private Google Access for any VM (not only for those who has an `internet` tag), but i have no clue how often those IP address ranges change, so I need to compare list pretty often.

**Comment by caieo** [comment]
@Atoms, I think I might see where the `nodeConfig` error is coming from. 

> 2. Create YAML with nodeConfig: tags: ['internet']
>>I use this in ContainerCluster as per documentation you cannot create ContainerCluster without nodes, therefore we need to create it with one node which is removed afterward and ContainerNodePool is attached to this ContainerCluster

When using `ContainerCluster` with the `remove-default-node-pool` annotation, you should *not* set a `nodeConfig` in the cluster config. You can set `initialNodeCount: 1`, but the initial node the cluster is spun up with will ignore/remove that default node. Because that node is ignored anyways, you shouldn't define a `nodeConfig` in `ContainerCluster` if you're using `ContainerNodePools`.  I would recommend setting up your cluster/nodepools the same way it is set up in the `ContainerNodePool` samples here: https://cloud.google.com/config-connector/docs/reference/resource-docs/container/containernodepool#sample_yamls

The `ContainerNodePool` attaches to the `ContainerCluster` via the `clusterRef` field (on `ContainerNodePool`). Regardless of which resource gets created first, Config Connector should be able to eventually reconcile the correctly tagged node pools to your cluster. Please try this out and let me know if that works!

**Comment by Atoms** [comment]
It would work if I would create those resources in async way, I'm waiting on resource to be ready, so if I create node pool before cluster I get it in status: 
```YAML
status:
  conditions:
  - lastTransitionTime: "2021-04-06T05:44:35Z"
    message: reference ContainerCluster project/cluster is not found
    reason: DependencyNotFound
    status: "False"
    type: Ready
```
and therefore my waiting on resources to become ready is not ending and the playbook is not getting to create a Container cluster.

So, I'm using ansible to deploy resources, and before I go to the next step in the playbook I want to make sure the previous step is successful, of course, I could skip here waiting on resource, but I always want to make sure resource is created (e.g. now I'm getting quota errors if i create bigger cluster, and if i would make it async i would not know that in playbook running stage)

**Comment by Atoms** [comment]
Just tested without waiting for cluster/nodepool to be ready, waited 36 minutes until I could use a cluster with my nodepool, and if I configure something wrong I need to go back to check after it's been created. 

So for now a workaround for adding routes to Private Google Access is the only way how to go around this issue. Also, I believe I will still need those routes for other resources too to access Private Google Access, so for now, i will go with that.

**Comment by Atoms** [comment]
And also I see that  `ContainerNodePool` nodeConfig and especially tags are immutable, which is somehow wrong, if i need adjust some tags for nodes in pool to allow/deny some firewall rules i need to recreate pool, which could lead to downtime cause applications already using this pool (i believe this need separate issue)

**Comment by caieo** [comment]
> It would work if I would create those resources in async way, I'm waiting on resource to be ready, so if I create node pool before cluster I get it in status: `[...]` and therefore my waiting on resources to become ready is not ending and the playbook is not getting to create a Container cluster.
> Just tested without waiting for cluster/nodepool to be ready, waited 36 minutes until I could use a cluster with my nodepool, and if I configure something wrong I need to go back to check after it's been created. 

I'm not that familiar with how Ansible works, but I think you should create the cluster before the nodePool if you don't want to get stuck on that status error. I feel like that is what you probably have been doing, but wasn't sure. I know that you mentioned you were unsure about the cluster being created with the default node pool and having to set the nodeConfig, but I believe once that `remove-default-node-pool` annotation is set, the cluster should ignore the initial node pool and then just connect to the `ContainerNodePool`'s config when its ready. 

Given all this information, can you clarify the bug you'd like us to address? I understand you found a workaround, but let us know if it would help to improve our documentation/add a sample on cluster routing/etc.

**Comment by Atoms** [comment]
In ansible there are 2 tasks which are needed for resource in K8s to be created:
1) apply templates resource (like container cluster)
2) wait till resource conditions.type == Ready and conditions.status == "True"

And that's how it's created for each resource type. So my flow is:
1) Create `ContainerCluster` with initialNodeCount: 1 and remove-default-node-pool annotation. 
2) Wait till `ContainerCluster` is ready
3) Add `ContainerNodePool` with clusterRef
4) Wait till 'ContainerNodePool` is ready

The issue is, that when `enablePrivateNodes` is true, and we don't have CloudNat for nodes (without default route 0.0.0.0/0 next-hop internet gateway), then nodes are unable to connect to Google Private endpoints. 

In my case I have CloudNat, but for nodes with network tag 'internet', when initial node does not have tag internet it cannot be provisioned, cause it cannot join the cluster, therefore cluster gets deleted and recreated after a timeout. I don't get Cluster is ready status from it to attach `ContainerNodePool` afterwards.

As I see it, the admission webhook should not deny updates if nodeconfig I'm applying is the same as it is already applied. If values change, then it blocks updates. That way I could add `nodeConfig: tags: ['internet']` and if I don't change that in the next playbook run it would not deny changes (only if I change `nodeConfig` values, then it can deny update)

This is not an issue when:
* Nodes have CloudNat - default route to the internet (also for Google Private Access which is in 0.0.0.0/0 route)
* If nodes have an external IP address.
* There is ~123 routes that route Google Private Access to the default internet gateway and nodes are still not able to reach the internet

This is an issue when:
* enablePrivateNodes is true and nodes do not have default internet route (0.0.0.0/0) 

I know there is a way to just apply and forget, and eventually, it will get to ready state, but that does not work in my planned workflow cause my next tasks are dependant on previous ones, so I need to wait until resource is ready.

**Comment by arekkusu** [comment]
> As I see it, the admission webhook should not deny updates if nodeconfig I'm applying is the same as it is already applied.

This happens because the *nodeConfig* field "will be populated with the node configuration of the first node pool."  Ref: [GKE API](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/projects.locations.clusters#Cluster)

So *nodeConfig* will initially reflect the default node pool but then when that gets deleted, it then reflect the first create Nodepool (*ContainerNodePool*). I know this is somewhat confusing / problematic, but it is how the GKE API is defined.

**Comment by Atoms** [comment]
So there is a problem to specify config for initialnode, so if my node cannot connect by default to Google Private API then i'm screwed?

I don't see actually that in GKE API nodeconfig would be immutable. 

So my option is to allow all nodes to connect to the internet with the default route by default, which means I'm adding a default route for all Instances also in that VPC.  What if i don't add CloudNat with default route to the private cluster? it is unable to connect to Google Private API then?
